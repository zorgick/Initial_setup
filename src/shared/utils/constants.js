export const WINDOW_STORE = '__preloaded_store__'
