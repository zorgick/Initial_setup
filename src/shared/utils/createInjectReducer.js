import * as React from 'react'
import storeManager from 'shared/configureStore'

const store = storeManager.getStore()
/**
 * This function is a sort of a middleware - it serves to join the
 * logic behind HOC and async redux store. <br />
 * GOAL: dynamically update a global redux store by injecting async reducers
 * @param {Function} createReducer function that injects async reducers.
 * @see [Twitter's approach](http://nicolasgallagher.com/redux-modules-and-code-splitting/)
 * @see [Abramov's approach](https://stackoverflow.com/questions/32968016/how-to-dynamically-load-reducers-for-code-splitting-in-a-redux-application)
 * @returns {Function} HOF
 * @example <caption>Step 1:</caption>
 * // `createInjectReducer` is called only once per runtime with createReducer:
 * export default createInjectReducer(createReducer)
 *
 * @example <caption>Step 2:</caption>
 * injectReducer({key: 'active', reducer})
 *
 * @example <caption>Step 3:</caption>
 * export default withReducer(Active)
 */

export default function createInjectReducer(createReducer) {
  return ({ key, reducer }) => Component =>
    class WithReducer extends React.PureComponent {
      constructor(props) {
        super(props)

        /**
         * Check the existance of the provided **key** ([[SCOPE]]) in the asyncReducers dictionary
         * and compare its value with the provided **reducer** ([[SCOPE]])
         * if not -> add the **reducer** to the dictionary
         */
        if (!(store.asyncReducers[key] && store.asyncReducers[key] === reducer)) {
          store.asyncReducers[key] = reducer

          /**
           * Update global store with *asyncReducers*.
           */

          store.replaceReducer(createReducer(store.asyncReducers))
        }
      }

      static displayName = `withReducer(${Component.displayName ||
        Component.name ||
        'Component'})`

      render() {
        return <Component {...this.props} />
      }
    }
}
