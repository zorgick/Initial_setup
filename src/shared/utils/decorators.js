export const debounce = (func, ms) => {
  let callStatus = false

  return (...rest) => {
    if (callStatus) {
      return
    }
    callStatus = true

    Reflect.apply(func, this, rest)

    setTimeout(() => {
      callStatus = false
    }, ms)
  }
}
