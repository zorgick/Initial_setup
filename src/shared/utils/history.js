import { createBrowserHistory, createMemoryHistory } from 'history'

export const createHistory = ({ initialEntries = [] } = {}) =>
  __BROWSER__ ? createBrowserHistory() : createMemoryHistory({ initialEntries })

export default createHistory
