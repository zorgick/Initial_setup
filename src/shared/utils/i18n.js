import fetch from 'isomorphic-fetch'
import i18next from 'i18next'
import { initReactI18next } from 'react-i18next'
import i18nFetch from 'i18next-fetch-backend'

import { RU, EN } from 'shared/i18n/allTranslations'

const { NODE_ENV, COMMIT = '' } = process.env
const cacheKey = COMMIT.slice(0, 20)

const backendOpts = {
  loadPath: `/locales/{{lng}}/{{ns}}.json${cacheKey ? '?' + cacheKey : ''}`,
  fetch
}

if (__BROWSER__) {
  i18next.use(i18nFetch)
}

i18next.use(initReactI18next).init({
  backend: backendOpts,
  react: {
    useSuspense: false,
    wait: true,
    bindI18n: 'languageChanged loaded',
    bindStore: 'added removed',
    nsMode: 'default'
  },
  load: 'languageOnly',
  debug: NODE_ENV === 'development' && __BROWSER__,
  fallbackLng: 'ru',
  fallbackNS: ['translation'],
  // Load missing resources via i18next-fetch-backend
  partialBundledLanguages: true,
  resources: {
    ru: { translation: RU.translation },
    en: { translation: EN.translation }
  },
  interpolation: {
    escapeValue: false
  },
  parseMissingKeyHandler: missing => {
    if (NODE_ENV === 'development' && __BROWSER__) {
      console.warn('MISSING TRANSLATION:', missing)
    }
    return missing
  }
})
export default i18next
