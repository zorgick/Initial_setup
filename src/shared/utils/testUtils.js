export { mountWrap, shallowWrap } from 'shared/utils/testUtilities/contextWrap'
export { translateMock } from 'shared/utils/testUtilities/translateMock'
