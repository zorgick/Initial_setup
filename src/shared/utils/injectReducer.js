import createReducer from 'shared/reducers'
import createInjectReducer from './createInjectReducer'

export default createInjectReducer(createReducer)
