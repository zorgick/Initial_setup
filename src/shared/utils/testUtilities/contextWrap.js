import { shallow, mount } from 'enzyme'
import history from 'shared/utils/history'
import { shape } from 'prop-types'

const router = {
  history,
  route: {
    location: {},
    match: {}
  }
}

const createContext = () => ({
  context: { router },
  childContextTypes: { router: shape({}) }
})

export function mountWrap (node) {
  return mount(node, createContext())
}

export function shallowWrap (node) {
  return shallow(node, createContext())
}
