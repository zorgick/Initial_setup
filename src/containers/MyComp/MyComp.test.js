import React from 'react'
import {MyComp} from './index'
import {shallowWrap, mountWrap} from 'utils/contextWrap'

describe('Testing MyComp container...', () => {
  let props
  const wrappedShallow = newProps =>
    shallowWrap(<MyComp {...props} {...newProps} />)

  const wrappedMount = newProps =>
    mountWrap(<MyComp {...props} {...newProps} />)

  beforeEach(() => {
    props = {}
  })

  test('should render MyComp container', () => {
    const wrapper = wrappedShallow()
    expect(wrapper).toMatchSnapshot()
  })
})
