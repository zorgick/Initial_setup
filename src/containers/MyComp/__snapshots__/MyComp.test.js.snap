// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing MyComp container... should render MyComp container 1`] = `
<div>
  <Link
    to="/"
  >
    back
  </Link>
</div>
`;
